columnar/src/columnar/merge.rs

```diff
@@ -31,3 +35,135 @@ pub fn merge_columnar(
         }
     }
 }
+
+pub fn collect_columns(
+    columnar_readers: &[&ColumnarReader],
+) -> io::Result<HashMap<String, HashMap<ColumnTypeCategory, Vec<DynamicColumn>>>> {
+    // Each column name may have multiple types of column associated.
+    // For merging we are interested in the same column type category since they can be merged.
+    let mut field_name_to_group: HashMap<String, HashMap<ColumnTypeCategory, Vec<DynamicColumn>>> =
+        HashMap::new();
+
+    for columnar_reader in columnar_readers {
+        let column_name_and_handle = columnar_reader.list_columns()?;
+        for (column_name, handle) in column_name_and_handle {
+            let column_type_to_handles = field_name_to_group
+                .entry(column_name.to_string())
+                .or_default();
+
+            let columns = column_type_to_handles
+                .entry(handle.column_type().column_type_category())
+                .or_default();
+            columns.push(handle.open()?);
+        }
+    }
+
+    normalize_columns(&mut field_name_to_group);
+
+    Ok(field_name_to_group)
+}
+
+/// Cast numerical type columns to the same type
+pub(crate) fn normalize_columns(
+    map: &mut HashMap<String, HashMap<ColumnTypeCategory, Vec<DynamicColumn>>>,
+) {
+    for (_field_name, type_category_to_columns) in map.iter_mut() {
+        for (type_category, columns) in type_category_to_columns {
+            if type_category == &ColumnTypeCategory::Numerical {
+                let casted_columns = cast_to_common_numerical_column(&columns);
+                *columns = casted_columns;
+            }
+        }
+    }
+}
+
+/// Receives a list of columns of numerical types (u64, i64, f64)
+///
+/// Returns a list of `DynamicColumn` which are all of the same numerical type
+fn cast_to_common_numerical_column(columns: &[DynamicColumn]) -> Vec<DynamicColumn> {
+    assert!(columns.iter().all(|column| column.is_numerical()));
+    let coerce_to_i64: Vec<_> = columns
```

Comment:
we can probably do something a little bit nicer here, and be more coupled with the writer logic?

```rust
fn acceptable_numerical_type(numerical_types: &[NumericalType]) -> NumericalType {
     let mut compatible_numerical_types = CompatibleNumericalTypes::default();
     for &typ in numerical_types {
         compatible_numerical_types.accept(typ);
     }
     numerical_types.to_numeric_type()
}


struct DynamicColummn {
    fn column_type(&self) -> ColumnType { ... }
    fn coerce(&self, column_type: ColumnType) -> Option<DynamicColumn> { ... }
}
```

Reply:
the type is not enough, we need to know if the a i64 column could be a u64 column (!contains negative value)

Reply:
ah good point